lib/Mojo/SMTP/Client.pm

```diff
@@ -385,28 +386,48 @@ sub _make_cmd_steps {
 					}
 					
 					my $data = $cmd[$mi]->();
+					my $data_buffer = ref $data ? $$data : $data;
 					
-					unless (length(ref $data ? $$data : $data) > 0) {
+					unless (length($data_buffer) > 0) {
 						$self->_write_cmd(($was_nl ? '' : Mojo::SMTP::Client::Response::CRLF).'.', CMD_DATA_END);
 						$self->_read_response($data_writer_cb, !$prepend && $mi == $#cmd);
 						$self->{expected_code} = CMD_OK;
 						return delete($self->{cleanup_cb})->();
 					}
-					
-					$was_nl = _has_nl($data);
-					$self->{stream}->write(ref $data ? $$data : $data, $data_writer);
+					# The following part if heavily inspired by Net::Cmd
+					my $first_ch = '';
+					# We have not send anything yet, so last_ch = "\012" means we are at the start of a line
+					$last_ch = "\012" unless defined $last_ch;
+					if ($last_ch eq "\015") {
+						# Remove \012 so it does not get prefixed with another \015 below
+						# and escape the . if there is one following it because the fixup
+						# below will not find it
+						$first_ch = "\012" if $data_buffer =~ s/^\012(\.?)/$1$1/;
+					}
+					elsif ($last_ch eq "\012") {
+						# Fixup below will not find the . as the first character of the buffer
+						$first_ch = "." if $data_buffer =~ /^\./;
+					}
+					$data_buffer =~ s/\015?\012(\.?)/\015\012$1$1/sg;
+					substr($data_buffer, 0, 0) = $first_ch;
+					$last_ch = substr($data_buffer, -1, 1);
+					$was_nl = _has_nl($data_buffer);
+					$self->{stream}->write($data_buffer, $data_writer);
 				};
 				
 				push @steps, $data_writer, $self->{resp_checker};
 			}
 			else {
+				my $data_buffer;
 				push @steps, sub {
 					my $delay = shift;
-					$self->{stream}->write(ref $cmd[$mi] ? ${$cmd[$mi]} : $cmd[$mi], $delay->begin);
+					$data_buffer = ref $cmd[$mi] ? ${$cmd[$mi]} : $cmd[$mi];
+					$data_buffer =~ s/\015?\012(\.?)/\015\012$1$1/sg; # turn . into .. if it's first character of the line
```

Comment:
And how about possible dot at the very beginning of the message? Should we also handle  it?